src/parser/core/modules/Gauge/CounterGauge.ts

```diff
@@ -66,25 +87,81 @@ export class CounterGauge extends AbstractGauge {
 	reset() {
 		// NOTE: This assumes counters always reset to their minimum value.
 		// Should that not be the case, probbaly needs a `resetTo` value.
-		this.set(this.minimum)
+		this.set(this.minimum, 'reset')
+	}
+
+	raise() {
+		this.set(this.minimum, 'init')
+	}
+
+	init() {
+		// Ensure we have a gauge init event, can't do in constructor because the parser reference might not be there yet
+		if (this.history.length === 0) {
+			this.history.push({
+				timestamp: this.parser.pull.timestamp,
+				value: this.initialValue,
+				minimum: this.minimum,
+				maximum: this.maximum,
+				delta: 0,
+				reason: 'init',
+			})
+		}
 	}
 
 	/** Modify the current value by the provided amount. Equivalent to `set(currentValue + amount)` */
 	modify(amount: number) {
 		this.set(this._value + amount)
 	}
 
+	/** Increase the current value by the provided amount. */
+	generate = (amount: number) => this.modify(amount)
+
+	/** Decrease the current value by the provided amount. */
+	spend = (amount: number) => this.modify(-amount)
+
 	/** Set the current value of the gauge. Value will automatically be bounded to valid values. Value over the maximum will be tracked as overcap. */
-	set(value: number) {
-		this._value = Math.min(Math.max(value, this.minimum), this.maximum)
+	set(value: number, reason?: GaugeEventReason) {
+		const delta = Math.abs(value - this.value)
+		if (delta === 0 && reason != null) { return }
+
+		if (reason == null) {
+			reason = delta > 0 ? 'generate' : 'spend'
+		}
 
-		// TODO: underflow means tracking was out of sync - look into backtracking to adjust history?
-		const diff = value - this._value
+		const newValue = Math.min(Math.max(value, this.minimum), this.maximum)
+
+		const diff = value - newValue
 		if (diff > 0) {
 			this.overCap += diff
+		} else if (diff < 0 && this.correctHistory) {
+			this.correctGaugeHistory(delta, this._value)
+		}
+
+		this._value = newValue
+		this.pushHistory(reason, delta)
+	}
+
+	private correctGaugeHistory(spenderCost: number, currentGauge: number) {
+		// Get the last generation or initialisation event we've recorded
+		const lastGeneratorIndex = _.findLastIndex(this.history, event => event.reason === 'generate' || event.reason === 'init')
```

Comment:
This approach is pretty specific to non-deterministic gauges; for a deterministic gauge it would be more correct to backtrack to the most recent init.  Could we reasonably give consumers an additional configuration option to choose which backtrack approach they want?  Otherwise you're going to get really funky looking gauge graphs if someone has some carry forward but not enough to immediately spend (e.g. a RDM comes in at 60 | 60 and builds up to a melee combo naturally before spending).

Reply:
Obviously I can't fully test the 'correcting but it's deterministic case' since DNC is most definitely not, but I've added the checks and logic for it in finding the last 'generator' index, and it's still working for DNC so... fingers crossed?